Embeddings directory structure

shntnu opened the pull request:
Related: https://github.com/jump-cellpainting/aws/issues/85#issuecomment-1245809221

Arkkienkeli commented on folder_structure.md:
````diff
@@ -254,6 +255,28 @@ The files are all produced by the [profiling-recipe](https://github.com/jump-cel
+The structure here is very similar to `analysis`
+
+```
+└── embeddings
````
Probably the first question to ask: is this folder structure inherited from some other software?

Arkkienkeli replied:
Also, do you know if Pycytominer follows similar folder structure? @shntnu

shntnu replied:
It's inherited from our past selves :) I've adapted it from the way we do things in https://cytomining.github.io/profiling-handbook/

Arkkienkeli commented on folder_structure.md:
```diff
@@ -254,6 +255,28 @@ The files are all produced by the [profiling-recipe](https://github.com/jump-cel
+└── embeddings
+    ├── 2021_04_26_Batch1
+    │   ├── BR00117035
+    │   │   └── efficientnet_v2_imagenet1k_s_feature_vector_2_ec756ff
```
I am wondering if it would make more sense to move model identifier on top of the structure. I understand the intention that if someone downloads the data related to plate, all types of embeddings will be included, though I am not sure if users usually have the same intuition. What do you think @shntnu ?

shntnu replied:
Good point

I'd favor the `embeddings` > `batch` > `plate` hierarchy because that's what we use for other folders.

But I see your point – `embeddings` > `efficientnet_v2_imagenet1k_s_feature_vector_ec756ff ` > `batch` > `plate` will make a lot more sense. 

The equivalent for CellProfiler would have been:

`analysis` > `cellprofiler_v4.2.4_ce356fe` > `batch` > `plate`, where 
- `ce356fe` is the hash of the analysis.cppipe
- `v4.2.4` is the version of CellProfiler

At present, we use `analysis` > `batch` > `plate`, and don't have any mechanism to nest versions.

If I follow my advice, then we'd use: 

`analysis` > `batch` > `plate` > `cellprofiler_v4.2.4_ce356fe` 

Hm – I am torn here because preserving this hierarchy has helped keep this systematized.

Unless you strongly object, I propose we proceed with what I've suggested here. 

>  I understand the intention that if someone downloads the data related to plate, all types of embeddings will be included, though I am not sure if users usually have the same intuition

Indeed, this is definitely unexpected but we will provide instructions for download that will help resolve this.


```sh
BATCH=TAORF
PLATES=plates.txt # list of plates
MODEL=efficientnet_v2_imagenet1k_s_feature_vector_ec756ff

parallel -a ${PLATES} \
  mkdir -p \
  cpg0019-moshkov-deepprofiler/broad/workspace/embeddings/${BATCH}/{1}/${MODEL}/ 

parallel -a ${PLATES} \
  aws s3 sync --no-sign-request \
  s3://cellpainting-gallery/cpg0019-moshkov-deepprofiler/broad/workspace/embeddings/${BATCH}/{1}/${MODEL}/ \
  cpg0019-moshkov-deepprofiler/broad/workspace/embeddings/${BATCH}/{1}/${MODEL}/ 
```

Arkkienkeli replied:
@shntnu 
I don't have objections regarding storage, though the question is if we aim to align DeepProfiler output, downstream analysis notebooks and this structure. 

If not, I could organize embeddings for sharing only as you are proposing.
If yes, well, it depends how far we want to go:
1) Modify DeepProfiler output and downstream analysis notebooks in the repository. 
2) Modify only downstream analysis notebooks in the repository.
3) Don't change anything in DeepProfiler and add a version of notebooks for this folder structure in the buckets. 
Personally I lean towards the option 3.

shntnu replied:
Thanks for laying this out, @Arkkienkeli 

Do I understand correctly that if we change the folder structure to  `embeddings` > `efficientnet_v2_imagenet1k_s_feature_vector_2_ec756ff ` > `batch` > `plate`, then you won't need to change anything (DeepProfiler nor notebooks), and also not need to create a different version of notebooks?

Arkkienkeli replied:
@shntnu 
> Do I understand correctly that if we change the folder structure to  embeddings > efficientnet_v2_imagenet1k_s_feature_vector_2_ec756ff > batch > plate, then you won't need to change anything (DeepProfiler nor notebooks), and also not need to create a different version of notebooks?

Yes, it will mean that DeepProfiler's output and the input of the aggregation notebook will be a single batch.

> In other words, if we do what you suggested in https://github.com/broadinstitute/cellpainting-gallery/pull/19#discussion_r986012440, then you are all set?

Yes.

Arkkienkeli replied:
@shntnu 
Let's just do the following: I will follow the already proposed structure as is to the possible extent and provide a modified version of aggregation Jupyter notebook to read it in `/workspace/analysis/` (option 3 from my comment above). 
No changes would be required here. I don't think we need to push the bucket's folder structure and\or the codebase to be unified.

shntnu replied:
@Arkkienkeli If this is not late, let's proceed with what you had initially proposed, i.e., change the folder structure to 

embeddings > efficientnet_v2_imagenet1k_s_feature_vector_2_ec756ff > batch > plate

shntnu replied:
I've made that change to the folder structure

Arkkienkeli commented on folder_structure.md:
```diff
@@ -254,6 +255,28 @@ The files are all produced by the [profiling-recipe](https://github.com/jump-cel
+    ├── 2021_04_26_Batch1
+    │   ├── BR00117035
+    │   │   └── efficientnet_v2_imagenet1k_s_feature_vector_2_ec756ff
+    │   │       ├── BR00117035-A01-1
```
Is not it redundant to have plate id here, or it is implied that someone might want to download site embedding separately?

shntnu replied:
Here, too, I am following what we have used in the past with CellProfiler, but indeed, we can skip the plateid prefix here. I will make that change

Arkkienkeli commented on folder_structure.md:
```diff
@@ -254,6 +255,28 @@ The files are all produced by the [profiling-recipe](https://github.com/jump-cel
+
+In this example batch:
+- `2021_04_26_Batch1` is the batch and `BR00117035` is the plate
+- `efficientnet_v2_imagenet1k_s_feature_vector_2_ec756ff` is an identifier for the deep learning network, prefixed with some hash for the model
```
Is the hash a necessary part of the name?

shntnu replied:
The goal here is to come up with a unique ID for the model. I'd be just as happy with a DOI (10.5281/zenodo.7114558 for CellPainting CNN v1)

Arkkienkeli replied:
Sounds good!

ErinWeisbart commented:
@Arkkienkeli A couple questions that I'm happy integrate into the documentation before merging:
- For someone with only basic knowledge of deep learning, can you describe in 1 sentence each what is found in the `collated`, `consensus`, and `embeddings` folders?
- For `efficientnet_v2_imagenet1k_s_feature_vector_2_ec756ff` which part of the string is an identifier for the deep learning network and which part the hash for the model?

For @shntnu 
- Must the various `.parquet` files be `.parquet` or do we allow other file types? .csv.gz? .sqlite?

Arkkienkeli commented:
@ErinWeisbart 
`collated`  - contains csv or parquet files with well-level profiles for all plates.
`consensus` - contains csv or parquet files with treatment-level profiles.

In this example, `efficientnet_v2_imagenet1k_s_feature_vector_2_ec756ff` -> `efficientnet` - name of the network (EfficientNet), `imagenet1k` - the dataset that was used for training, `ec756ff` - hash. 
We also agreed with @shntnu that it is possible to use other identifiers for model, for example zenodo doi (see [folder structure for DeepProfiler](https://github.com/broadinstitute/cellpainting-gallery/issues/20#issuecomment-1285787463)).

ErinWeisbart commented:
Thanks so much Nikita. 
I've moved things around a little bit and added some additional information to the documentation based on your help.
@Arkkienkeli and @shntnu would you take a look to make sure that what I've added is correct? Then I'm happy to have this merged.

Arkkienkeli commented:
@ErinWeisbart Looks good to me! Thank you.

shntnu commented:
@ErinWeisbart LGTM!

Thanks for the edits :)

shntnu commented:
> I've moved things around a little bit and added some additional information to the documentation based on your help.
> @Arkkienkeli and @shntnu would you take a look to make sure that what I've added is correct? Then I'm happy to have this merged.

@ErinWeisbart @Arkkienkeli many thanks! Merging now
